Assets/Scripts/ProceduralGeneration.cs

```diff
@@ -0,0 +1,192 @@
+    public int chunkSize;
+    public int unitCubeSize;
+    public float asteroidDensity = 0.5f;
+    public Vector2 asteroidSizeRange;
+    public float viewDistance = 500f;
+    
+    // public GameObject blackHolePrefab;
+    // public GameObject exitPortalPrefab;
+    public GameObject asteroidPrefab;
+
+    private GameObject asteroidsCollection;
+    
+    private Vector3Int numChunksInLevel;
+    private Vector3Int numUnitCubesInChunk;
+
+    private CullingGroup cullGroup;
+    private BoundingSphere[] bounds;
+    private Dictionary<int, GameObject> entityLookup;
+    private int EntityCount => entityLookup.Count;
+
+    private Dictionary<int, Transform> entityTransforms;
+    
+    private Dictionary<int, int> entityIdToIndex;
+    private Dictionary<int, int> entityIndexToId;
+    
+    public static bool FinishedGenerating = false;
+    private int maxEntities;
+
+    private void Awake()
+    {
+        asteroidsCollection = GameObject.FindGameObjectWithTag("Asteroids");
+
+        numChunksInLevel = levelDimensions / chunkSize;
+        numUnitCubesInChunk = Vector3Int.one * (chunkSize / unitCubeSize);
+
+        // var asteroidsPerChunk = 1; TODO: Make this more extendable
+        maxEntities = numChunksInLevel.x * numUnitCubesInChunk.x * numChunksInLevel.y * numUnitCubesInChunk.y * numChunksInLevel.z * numUnitCubesInChunk.z  * 1; // 1 asteroid per unit cube right now
+        Debug.Log($"Max Entities: {maxEntities}");
+        // Culling Setup
+        var mainCam = Camera.main;
+        cullGroup = new CullingGroup {targetCamera = mainCam};
+        cullGroup.SetDistanceReferencePoint(mainCam.transform);
+        cullGroup.SetBoundingDistances(new[] {viewDistance, float.PositiveInfinity});
+        bounds = new BoundingSphere[maxEntities];
+        cullGroup.SetBoundingSpheres(bounds);
+        cullGroup.SetBoundingSphereCount(0);
+        cullGroup.onStateChanged += CullingStateChanged;
+        entityLookup = new Dictionary<int, GameObject>();
+        entityTransforms = new Dictionary<int, Transform>();
+        entityIdToIndex = new Dictionary<int, int>();
+        entityIndexToId = new Dictionary<int, int>();
+
+        StartCoroutine(nameof(GenerateAsteroidField));
+        StartCoroutine(nameof(UpdateCulledObjectBounds));
+    }
+
+    public void AddEntity(GameObject obj)
+    {
+        var id = obj.GetInstanceID();
+        var tf = obj.transform;
+        var radius = obj.GetComponent<Renderer>().bounds.size.x;
+        entityLookup.Add(id, obj);
+        var index = EntityCount - 1;
+        bounds[index] = new BoundingSphere {position = tf.position, radius = radius};
+        entityTransforms.Add(id, tf);
+        entityIndexToId.Add(index, id);
+        entityIdToIndex.Add(id, index);
+        cullGroup.SetBoundingSphereCount(EntityCount);
+    }
+
+    public void RemoveEntity(GameObject obj)
+    {
+        var id = obj.GetInstanceID();
+        var index = entityIdToIndex[id];
```

Comment:
While play testing I got the following error multiple times (consistently with subsequent game sessions) as the black hole consumes stuff:

```KeyNotFoundException: The given key was not present in the dictionary.
System.Collections.Generic.Dictionary`2[TKey,TValue].get_Item (TKey key) (at <fb001e01371b4adca20013e0ac763896>:0)
ProceduralGeneration.RemoveEntity (UnityEngine.GameObject obj) (at Assets/Scripts/ProceduralGeneration.cs:86)
BlackHoleConsumption.OnTriggerEnter (UnityEngine.Collider other) (at Assets/Scripts/BlackHoleConsumption.cs:29)

Reply:
This is the transient error I mentioned.......

Reply:
I think this might be due to the black hole thinking it needs to delete something that's already been deleted...? Weird af